### PR TITLE
[llama-agents] Add an example w/ llama-agents and Milvus

### PR DESCRIPTION
**Pull request by stephen37**
This PR adds an example on how to use llama-agents with Milvus. It showcases the ability to use an agent to make decision based on the user's query. 
In this example, we show that the agent can pick which Milvus collection to use.

### CODE REVIEW

File: bootcamp/tutorials/integration/llama_agents_milvus.ipynb [hunk 1]
```diff
@@ -0,0 +1,569 @@
+{
```
Comment (codingjaguar) [inline bootcamp/tutorials/integration/llama_agents_milvus.ipynb:1]: <p> This repo contains the core components of Milvus Lite.</p><p>-- this is not related to this notebook.</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/milvus-io/bootcamp/pull/1401/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (codingjaguar) [inline bootcamp/tutorials/integration/llama_agents_milvus.ipynb:1]: <p>nit, the logo is a bit too large.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/milvus-io/bootcamp/pull/1401/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (codingjaguar) [inline bootcamp/tutorials/integration/llama_agents_milvus.ipynb:1]: <p>similar to previous comment, these code blocks are a bit too dry. there shall be some lubricates like illustration "in the next section we need to blabla"</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/milvus-io/bootcamp/pull/1401/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

File: bootcamp/tutorials/integration/llama_agents_metadata.ipynb [hunk 2]
```diff
@@ -0,0 +1,767 @@
+{
```
Comment (codingjaguar) [inline bootcamp/tutorials/integration/llama_agents_metadata.ipynb:1]: <p>Nit, the logo seems too big. Should be able to scale it in notebook?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/milvus-io/bootcamp/pull/1401/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (codingjaguar) [inline bootcamp/tutorials/integration/llama_agents_metadata.ipynb:1]: <p>the visual is very intuitive. good point.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/milvus-io/bootcamp/pull/1401/discussion/#comment-1705463975'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

Comment (codingjaguar) [inline bootcamp/tutorials/integration/llama_agents_metadata.ipynb:1]: <p>worth explaining what these few lines about load dotenv is doing.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/milvus-io/bootcamp/pull/1401/discussion/#comment-1705463979'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

Comment (codingjaguar) [inline bootcamp/tutorials/integration/llama_agents_metadata.ipynb:1]: <p>this part is no longer download data.</p><p> </p><p>something simple like:</p><p> </p><p>Prepare Embedding Model</p><p> </p><p>We use open-source embedding model xxx listed on Hugging Face xxxx</p><p>-- code --</p><p> </p><p>should do the job.</p><p> </p><p>Same idea applies to the following sections.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/milvus-io/bootcamp/pull/1401/discussion/#comment-1705463981'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

Comment (codingjaguar) [inline bootcamp/tutorials/integration/llama_agents_metadata.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #9.        uri="./milvus_demo_metadata.db",
</pre><p>nit, why "metadata", e.g. is this demo specifically about metadata storage or filtering? otherwise it feels a little unrelated.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/milvus-io/bootcamp/pull/1401/discussion/#comment-1705463983'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

Comment (codingjaguar) [inline bootcamp/tutorials/integration/llama_agents_metadata.ipynb:1]: <p>i think there could be a bit more explaination before this code block. e.g. emphasizing that here we need to define a few agent tools through natural language descriptions, and the tools are what the "agent service" blocks in the top graph is about. otherwise the user may have a hard time understanding how come a few description strings help in this story.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/milvus-io/bootcamp/pull/1401/discussion/#comment-1705463984'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

Comment (codingjaguar) [inline bootcamp/tutorials/integration/llama_agents_metadata.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #8.    agent = ReActAgent.from_tools(query_engine_tools, llm=llm, verbose=True)
</pre><p>may explaining what react is and give a link to it, <a href="https://www.promptingguide.ai/techniques/react" target="_blank">https://www.promptingguide.ai/techniques/react</a></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/milvus-io/bootcamp/pull/1401/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

Comment (codingjaguar) [inline bootcamp/tutorials/integration/llama_agents_metadata.ipynb:1]: <p>this block isn't straightforward. worth explaining at least two things:</p><p>1) what does the metadata filtering has to do with agent tools, how are those two combined to achieve some objective</p><p>2) what's the difference/relationship between this QueryEngineTool to the previews uber and lyft tools. (actually i didn't get it from reading it the first time)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/milvus-io/bootcamp/pull/1401/discussion/#comment-1705463989'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

Comment (codingjaguar) [inline bootcamp/tutorials/integration/llama_agents_metadata.ipynb:1]: <p>before this test, there shall be a "preview" of the story saying that the next one won't answer about uber as the filter only allows lyft related content to be retrieved. While the next qusetion can be answer as it's related to lyft. We don't have to let users crack the puzzle and guess that.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/milvus-io/bootcamp/pull/1401/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='18'/>

Comment (codingjaguar) [inline bootcamp/tutorials/integration/llama_agents_metadata.ipynb:1]: <p>this piece is the jewel of the story, which shows how to use an agent to extract metadata filter and apply that during question answering. so</p><p>1) this point can be mentioned at the very begin of this story.</p><p>2) this part is worth a subtitle or some kind of highlighting</p><p> </p><p>What could make this story even more shiny is to show an example that, without the auto-metadata-filtering, the question answering will confuse itself by answering a qusetion about lyft with information about uber. (if that's possible to find such a counter case)</p><p> </p><p>The other case above does show different result, but it's unclear to the reader whether it's wrong and why:</p><p> </p><p>Response without metadata filtering: </p><p>The total revenue for Uber in 2021 was $17,455 million. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/milvus-io/bootcamp/pull/1401/discussion/#comment-1705463994'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='20'/>

Comment (codingjaguar) [inline bootcamp/tutorials/integration/llama_agents_metadata.ipynb:1]: <p>what's special about the question being answered below? why is that so difficult and why "Orchestrating the different services with Mistral Large" would help in that case?</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/milvus-io/bootcamp/pull/1401/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='23'/>

Comment (stephen37) [reply bootcamp/tutorials/integration/llama_agents_metadata.ipynb:1]: <p>It kinda does, it scales with the size of your screen basically. as far as I know Markdown doesn't really support scaling, I could take a smaller image though I guess. </p>

Comment (stephen37) [reply bootcamp/tutorials/integration/llama_agents_metadata.ipynb:1]: <p>Yeah, it was because I had another demo that was running and the db was called <code>milvus_demo.db</code>,I'll change it.   </p>

[the remark this answers is not in the document]
Comment (stephen37) [reply bootcamp/tutorials/integration/llama_agents_metadata.ipynb:1]: <p>Fixed it :D</p>

Comment (stephen37) [reply bootcamp/tutorials/integration/llama_agents_metadata.ipynb:1]: <p>Good point, added some text and a link to it</p>

**Comment by stephen37** [comment]
@codingjaguar Ready for review again, thanks for the comments btw, it made the Notebook so much better! 💙

File: bootcamp/tutorials/integration/llama_agents_metadata.ipynb (hunk 2 above)
Comment (codingjaguar) [reply bootcamp/tutorials/integration/llama_agents_metadata.ipynb:1]: <p>now looks much better.</p>

File: bootcamp/tutorials/integration/llama_agents_metadata.ipynb [hunk 3]
```diff
@@ -0,0 +1,1130 @@
...
```
Comment (codingjaguar) [inline bootcamp/tutorials/integration/llama_agents_metadata.ipynb:1]: <p>Nit, if possible, a slightly deeper one-sentence comment can be more helpful, e.g. xxx is needed for printing ... in beautiful format. For this one, it might be nest_asyncio.apply() is needed so that the asyn calls later used for agent ... can work in notebook. you get the idea.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/milvus-io/bootcamp/pull/1401/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

Comment (codingjaguar) [inline bootcamp/tutorials/integration/llama_agents_metadata.ipynb:1]: <p>this is good to explain why using any particular model.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/milvus-io/bootcamp/pull/1401/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

Comment (codingjaguar) [inline bootcamp/tutorials/integration/llama_agents_metadata.ipynb:1]: <p>good to see that the template saves effort. let's continue polishing the templates.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/milvus-io/bootcamp/pull/1401/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='19'/>

Comment (codingjaguar) [inline bootcamp/tutorials/integration/llama_agents_metadata.ipynb:1]: <p>The above code block has good explanation, but this block misses such things. E.g. link to ReAct as i mentioned in another comment, and how the two tools are playing here and why missing metadata filtering is bad.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/milvus-io/bootcamp/pull/1401/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='23'/>

Comment (codingjaguar) [inline bootcamp/tutorials/integration/llama_agents_metadata.ipynb:1]: <p>Here is a problem. ReAct has already been used in previous code block, but only mentioned until here. It's already too late. May previous one using react is not your intention?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/milvus-io/bootcamp/pull/1401/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='26'/>

**Comment by stephen37** [comment]
@codingjaguar  I removed ReActAgent, it was a bit of a pain, I also addressed your comments, let me know what you think now 🙌

**Review by codingjaguar — APPROVED** [review]

File: bootcamp/tutorials/integration/llama_agents_metadata.ipynb [hunk 4]
```diff
@@ -0,0 +1,940 @@
...
```
Comment (codingjaguar) [inline bootcamp/tutorials/integration/llama_agents_metadata.ipynb:1]: <p>Nit, i think put</p><p> </p><h1>Goal of this Notebook </h1><p> </p><h1>🔍 Summary </h1><p> </p><p>ahead of introduction of Milvus and other tech is slightly better.</p><p> </p><p>That way there is a clear opening before diving into details.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/milvus-io/bootcamp/pull/1401/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (codingjaguar) [inline bootcamp/tutorials/integration/llama_agents_metadata.ipynb:1]: <p>remvoe "This repo contains the core components of Milvus Lite."?</p><p> </p><p>better have some transition, e.g. "Milvus is xxx. In this demo we use milvus lite, a lightweight version of milvus. Milvus Lite is xxx"</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/milvus-io/bootcamp/pull/1401/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (stephen37) [reply bootcamp/tutorials/integration/llama_agents_metadata.ipynb:1]: Moved it

**Review by codingjaguar — APPROVED** [review]
